WARP-5838: Separate driver into MDIO & platform modules

bolt-snap-bolt opened the pull request:
- Separate lantiq_gswip.c driver file out into 4 consituent files as first step
in adding support for MDIO-managed external parts.
- Keep majority of driver logic in lantiq_gsw_core.c and lantiq_gsw.h
- Isolate platform driver specific logic into lantiq_gsw_platform.c
- Add new lantiq_gsw_mdio.c file to hold MDIO-specific logic and annotate
with TODO statements where additional work is to be carried out
- Edit Kconfig to allow for granular configuration of platform and MDIO driver
- Edit Makefile to ensure new files are built
- This is a first pass at slicing up these files--changes are likely

bolt-snap-bolt committed:
WARP-5838: Separate driver into MDIO & platform modules

- Separate lantiq_gswip.c driver file out into 4 consituent files as first step
in adding support for MDIO-managed external parts.
- Keep majority of driver logic in lantiq_gsw_core.c and lantiq_gsw.h
- Isolate platform driver specific logic into lantiq_gsw_platform.c
- Add new lantiq_gsw_mdio.c file to hold MDIO-specific logic and annotate
with TODO statements where additional work is to be carried out
- Edit Kconfig to allow for granular configuration of platform and MDIO driver
- Edit Makefile to ensure new files are built
- This is a first pass at slicing up these files--changes are likely

avlec commented on drivers/net/dsa/lantiq_gsw.h:
```diff
@@ -0,0 +1,307 @@
+
+#define XRX200_GPHY_FW_ALIGN	(16 * 1024)
+
+struct gsw_hw_info {
```
No more is_external member the code in lantiq_gsw_core.c still references that member (line 2025/1747) is this getting compiled?

bolt-snap-bolt replied:
It was in fact _not_ getting compiled, due to some kconfig/Makefile issues.
PR will be updated in the coming days once I resolve the glut of compilation errors.

avlec commented on drivers/net/dsa/lantiq_gsw_platform.c:
```diff
@@ -0,0 +1,114 @@
+{
+	u32 val;
+
+	return readx_poll_timeout(__raw_readl, priv->gswip + (offset * 4), val,
```
This is from the original driver but use of uninitialized variable val?

mclarks replied:
readx_poll_timeout is a macro mapping to another macro read_poll_timeout defined in include/linux/iopoll.h. Looks like when it is evaluated, val is set before the condition (val & cleared) == 0.

mclarks commented on drivers/net/dsa/lantiq_gsw_mdio.c:
```diff
@@ -0,0 +1,124 @@
+
+static const struct of_device_id gsw_mdio_of_match[] = {
+	{ .compatible = "maxlinear,gsw120", .data = &gsw_120 },
+	{ .compatible = "maxlinear,gsw125", .data = &gsw_120 },
```
If gsw120 and gsw125 are essentially the same (just different temperature ranges) would is make sense to have a single entry which is gsw12x?

bolt-snap-bolt replied:
Since both map to the same data it doesn't really matter, but in the interest of not having to edit our device tree string I've changed this to a single "maxlinear,gsw12x" line 👍

bolt-snap-bolt committed:
WARP-5838: Separate driver into MDIO & platform modules
- Edit Kconfig changes to ensure core file is included as expected
- Move some functions back into core file and add read/write function
pointers to access mdio & platform specific functions.
- Misc. cleanup & edits to get it compiling successfully.

bolt-snap-bolt commented on drivers/net/dsa/lantiq_gsw_mdio.c:
```diff
@@ -0,0 +1,103 @@
+static int gsw_mdio_probe(struct mdio_device *pmdiodev)
+{
+	/* TODO WARP-5829: write this function */
+	printk("!RCC: GSW MDIO probe func invoked.\n");
```
I've confirmed I see these strings in the kernel message buffer on boot when my device tree has an MDIO-bus sub-node that specifies the new "maxlinear,gsw12x" compatible string.

bolt-snap-bolt commented on drivers/net/dsa/lantiq_gsw.h:
```diff
@@ -0,0 +1,309 @@
+	u32 port_vlan_filter;
+};
+
+struct gsw_ops {
```
I moved a lot of the functions back into the core file and have added this new struct to allow the core code to call bus-specific read/write functions.

avlec replied:
Big fan of this

bolt-snap-bolt commented on drivers/net/dsa/lantiq_gsw_core.c:
```diff
@@ -363,40 +118,42 @@ static void gswip_switch_mask(struct gswip_priv *priv, u32 clear, u32 set,
 static u32 gswip_switch_r_timeout(struct gswip_priv *priv, u32 offset,
```
The body of this function is currently commented out as I'll need to get far enough along to learn whether I can drop the `gswip_priv` argument from the `ops.read()` call. If I can, it will just call `priv->ops->read()` and stay here. If not, I will likely convert it into a bus-specific function and add it to the `ops` struct.

avlec replied:
Shouldn't you be able to, cause you defined how those functions are from the existing `gswip_switch_r`. Is there any other information in `gswip_priv` that needs to be known other than the address that gets calculated? The thing I can see is the address offset calculation needing to be done differently between the platform/mdio code.

bolt-snap-bolt replied:
The `readx_poll_timeout` macro expects a read function prototype with only a single argument (currently the offset). In the MDIO case there is additional needed information: a handle to the MDIO device in order to call the correct bus function, know what address on the bus to target, etc.

There are a couple options: move this function to the bus-specific files, bundle the necessary arguments into a struct and pass that, or make the MDIO handle / state info global and remove the need to pass it at all. Not sure exactly which approach I'll take yet...

avlec replied:
I am not a fan of the global option I think the other option sounds a lot better. You could also just add a function pointer for the timeout read variant? So you'd have `priv->ops->read(address)` and `priv->ops->r_timeout(address)`

bolt-snap-bolt replied:
> I am not a fan of the global option I think the other option sounds a lot better. You could also just add a function pointer for the timeout read variant? So you'd have `priv->ops->read(address)` and `priv->ops->r_timeout(address)`

Ended up going that route, yup. See https://github.com/bolt-snap-bolt/linux/compare/feature/WARP-5838...bolt-snap-bolt:feature/WARP-5840
